activestorage/app/models/active_storage/blob.rb

```diff
@@ -53,19 +53,21 @@ def find_signed(id)
     # When providing a content type, pass <tt>identify: false</tt> to bypass automatic content type inference.
     def build_after_upload(io:, filename:, content_type: nil, metadata: nil, identify: true)
       new(filename: filename, content_type: content_type, metadata: metadata).tap do |blob|
+        blob.reserve_key!
```

Comment:
I think that this makes the documentation for this method ("returns new, unsaved blob instance") incorrect and the method name misleading. 

I think that if the possible key collision is an issue, the `build_*` methods should be removed - and we should instead corresponding `create_*` methods.

I If we are creating before uploading I don't think that the `reserve_key!` method would still be required.

Reply:
> I think that if the possible key collision is an issue, the build_* methods should be removed - and we should instead corresponding create_* methods.

I would vote for that too. Don't know how much "public-ness" is attached to those build methods though, so not my call to make. "Create before upload always" is the proposed solution anyway.

Reply:
I think I’m okay with removing `build_after_upload`. `build_after_unfurling` will likely need to stay to preserve desirable behaviors of the attaching API, but it’s strictly internal and used safely.